src/crimson/os/seastore/journal/circular_bounded_journal.cc

```diff
@@ -0,0 +1,573 @@
+  ).safe_then(
+    [this, addr, read_length, bptr]() mutable
+    -> read_record_ret {
+      record_group_header_t h;
+      bufferlist bl;
+      bl.append(bptr);
+      auto bp = bl.cbegin();
+      try {
+	decode(h, bp);
+      } catch (ceph::buffer::error &e) {
+	return read_record_ret(
+	  read_record_ertr::ready_future_marker{},
+	  std::nullopt);
+      }
+      /*
+       * |          journal          |
+       *        | record 1 header |  | record 1 data
+       *  record 1 data  (remaining) |
+       *
+       *        <---- 1 block ----><--
+       * -- 2 block --->
+       *
+       *  If record has logner than read_length and its data is located across
+       *  the end of journal and the begining of journal, we need three reads
+       *  ---reads of header, other remaining data before the end, and
+       *  the other remaining data from the begining.
+       *
+       */
+      if (h.mdlength + h.dlength > read_length) {
+	rbm_abs_addr next_read_addr = addr + read_length;
+	auto next_read = h.mdlength + h.dlength - read_length;
+	logger().debug(" next_read_addr {}, next_read_length {} ",
+	    next_read_addr, next_read);
+	if (header.end < next_read_addr + next_read) {
+	  // In this case, need two more reads.
+	  // The first is to read remain bytes to the end of cbjournal
+	  // The second is to read the data at the begining of cbjournal
+	  next_read = header.end - (addr + read_length);
+	}
+	logger().debug("read_entry: additional reading addr {} length {}",
+			next_read_addr,
+			next_read);
+	auto next_bptr = bufferptr(ceph::buffer::create_page_aligned(next_read));
+	next_bptr.zero();
+	return device->read(
+	  next_read_addr,
+	  next_bptr
+	).safe_then(
+	  [this, h=h, next_bptr=std::move(next_bptr), bl=std::move(bl)]() mutable {
+	    bl.append(next_bptr);
+	    if (h.mdlength + h.dlength == bl.length()) {
+	      logger().debug("read_record: record length {} done", bl.length());
+	      return return_record(h, bl);
+	    }
+	    // need one more read
+	    auto next_read_addr = get_start_addr();
+	    auto last_bptr = bufferptr(ceph::buffer::create_page_aligned(
+		  h.mdlength + h.dlength - bl.length()));
+	    logger().debug("read_record: last additional reading addr {} length {}",
+			    next_read_addr,
+			    h.mdlength + h.dlength - bl.length());
+	    return device->read(
+	      next_read_addr,
+	      last_bptr
+	    ).safe_then(
+	      [this, h=h, last_bptr=std::move(last_bptr), bl=std::move(bl)]() mutable {
+		bl.append(last_bptr);
+		logger().debug("read_record: complte size {}", bl.length());
+		return return_record(h, bl);
+	      });
+	});
+      } else {
+	logger().debug("read_record: complte size {}", bl.length());
+	return return_record(h, bl);
+      }
+    });
+}
+
+CBJournal::write_ertr::future<>
+CBJournal::sync_super()
+{
+  ceph::bufferlist bl;
+  try {
+    bl = encode_super();
+  } catch (ceph::buffer::error &e) {
+    logger().debug("unable to encode super block from underlying deivce");
```

Comment:
This isn't a device problem, this can only happen if there's an actual software bug in the implementation of encode.